mutex.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package mutex
+
+import (
+	"regexp"
+	"time"
+
+	"github.com/juju/errors"
+)
+
+var (
+	validName = regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9-]*$")
+)
+
+// Releaser defines the Release method that is the only thing that can be done
+// to a acquired mutex.
+type Releaser interface {
+	// Release makes the mutex available for others and invalidates the instance.
+	// Subsequent calls to Release will error.
+	Release() error
+}
+
+// Acquirer defines the standard method to acquire a mutex.
+type Acquirer interface {
+	Acquire() (Releaser, error)
+}
+
+// Clock provides an interface for dealing with clocks.
+type Clock interface {
+	// After waits for the duration to elapse and then sends the
+	// current time on the returned channel.
+	After(time.Duration) <-chan time.Time
+}
+
+// Spec defines the name of the mutex and behaviour of the Acquire function.
+type Spec struct {
+	// Name is required, and must start with a letter and contain at most
+	// 40 letters, numbers or dashes.
+	Name string
+	// Clock must be provided and is exposed for testing purposes.
+	Clock Clock
+	// Delay defines how often to check for lock acquisition. The implementation
+	// calls all fail fast, so acquisition requires polling.
+	Delay time.Duration
+	// Timeout allows the caller to specify how long to wait.
+	Timeout time.Duration
+	// Cancel if signalled will cause the Acquire method to return with ErrCancelled.
+	Cancel <-chan struct{}
+}
+
+// Acquire implements Acquirer.
```

Comment:
@fwereade same here. One of the design decisions that thumper and I talked about was never having a Mutex which is not locked -- that is, if you have a value which implements `Releaser` your process holds the lock.

Currently a lot of the fslock code set's up the lock (to coin a phrase) then passes around that *fslock inside a config or a context or some other structure to the point that the calling code wants to grab the lock.

This has the downside that if your function receives an *fslock, you don't know if it's locked or not. And this is a big problem, because this lock is machine wide. Any operation like `IsLocked() bool` is inherently racy because between the time of asking "is it locked" and deciding "right, it's not locked, I'll go lock it" or worse, _assume_ that it is not locked while the code does something, the lock could be take by any other process on the machine.

The `mutex.Spec` in this type avoids this ambiguity as it is the "setup" for a lock, and only when you want to grab the lock to you pass the spec to `mutex.Acquire()` and either grab the lock or don't. That is to say, if you have a Spec, you don't hold the lock, all you have is a description of the lock you'd like to hold.

So, with that said, I'd prefer `mutex.Acquire(Spec)`, rather than `func (m *Mutex) Acquire()`.